### ISSUE TITLE
fzf  problem

### ISSUE THREAD

**Issue by wjljack** [fish-shell/fish-shell#8963]
I use in Mac OS 12.4 with 
- fish 3.4.1 (with PatrickF1/fzf.fish plugins)
- fzf 0.30.0
and in user home folder, press ctrl-t is ok ,but in other folder ,it said:`warning: failed to set up file descriptors in child_setup_process` and fzf found nothing (0/0)
this warning message printed by fish shell  source 
`https://github.com/search?q=failed+to+set+up+file+descriptors+in+child_setup_process&type=code`

**Comment by faho** [comment]
What is ctrl-t bound to?

What is your fzf configuration? Do you have $FZF_CTRL_T_COMMAND or $FZF_DEFAULT_OPTS set to anything?

Do you use tmux?

>and in user home folder, press ctrl-t is ok ,but in other folder 

Does it *always* work in your home directory and *always* fail in a specific other folder?

In general, what we want is to have some way to reproduce your problem, and what you've given there isn't enough. We need more information. Ideally, you would describe the minimum number of steps to get to your problem from a fresh configuration.

**Comment by wjljack** [comment]
and  i find ,with fish initial path, is ok 
```
#open iterm2 or terminal.app
~ $ fzf-file-widget //ok
~ $ cd ~/Downloads
~/Downloads $ fzf-file-widget //fail
 ~/Downloads $ fish //enter fish again
~/Downloads $ fzf-file-widget //ok
~/Downloads $ cd .. // go home again
~ $ fzf-file-widget //fail
```

**Comment by wjljack** [comment]
for the clean env
```
# install fish with pkg
# install fisher
curl -sL https://raw.githubusercontent.com/jorgebucaran/fisher/main/functions/fisher.fish | source && fisher install jorgebucaran/fisher
# install fzf
brew install fzf
# install fzf default binding
$(brew --prefix)/opt/fzf/install
# install fzf.fish
fisher install PatrickF1/fzf.fish

```

**Comment by faho** [comment]
Okay. What does `type fzf-file-widget` say?

And, for good measure, what about `type eval`?

**Comment by wjljack** [comment]
type fzf-file-widget
```bash
# Defined in /Users/wangjunlong/.config/fish/functions/fish_user_key_bindings.fish @ line 19
function fzf-file-widget --description 'List files and folders'
    set -l commandline (__fzf_parse_commandline)
    set -l dir $commandline[1]
    set -l fzf_query $commandline[2]
    set -l prefix $commandline[3]

    # "-path \$dir'*/\\.*'" matches hidden files/folders inside $dir but not
    # $dir itself, even if hidden.
    test -n "$FZF_CTRL_T_COMMAND"; or set -l FZF_CTRL_T_COMMAND "
    command find -L \$dir -mindepth 1 \\( -path \$dir'*/\\.*' -o -fstype 'sysfs' -o -fstype 'devfs' -o -fstype 'devtmpfs' \\) -prune \
    -o -type f -print \
    -o -type d -print \
    -o -type l -print 2> /dev/null | sed 's@^\./@@'"

    test -n "$FZF_TMUX_HEIGHT"; or set FZF_TMUX_HEIGHT 40%
    begin
      set -lx FZF_DEFAULT_OPTS "--height $FZF_TMUX_HEIGHT --reverse --bind=ctrl-z:ignore $FZF_DEFAULT_OPTS $FZF_CTRL_T_OPTS"
      eval "$FZF_CTRL_T_COMMAND | "(__fzfcmd)' -m --query "'$fzf_query'"' | while read -l r; set result $result $r; end
    end
    if [ -z "$result" ]
      commandline -f repaint
      return
    else
      # Remove last token from commandline.
      commandline -t ""
    end
    for i in $result
      commandline -it -- $prefix
      commandline -it -- (string escape $i)
      commandline -it -- ' '
    end
    commandline -f repaint

end
```

type eval

```bash
eval is a function with definition
# Defined in /usr/local/share/fish/functions/eval.fish @ line 1
function eval --description 'Evaluate parameters as a command' --no-scope-shadowing
    # keep a copy of the previous $status and use restore_status
    # to preserve the status in case the block that is evaluated
    # does not modify the status itself.
    set -l status_copy $status
    function __fish_restore_status
        return $argv[1]
    end

    if not set -q argv[2]
        # like most builtins, we only check for -h/--help
        # if we only have a single argument
        switch "$argv[1]"
            case -h --help
                __fish_print_help eval
                return 0
        end
    end

    # If we are in an interactive shell, eval should enable full
    # job control since it should behave like the real code was
    # executed.  If we don't do this, commands that expect to be
    # used interactively, like less, wont work using eval.

    set -l mode
    if status --is-interactive-job-control
        set mode interactive
    else
        if status --is-full-job-control
            set mode full
        else
            set mode none
        end
    end
    if status --is-interactive
        status --job-control full
    end
    __fish_restore_status $status_copy

    # To eval 'foo', we construct a block "begin ; foo; end <&3 3<&-"
    # Note the redirections are also within the quotes.
    #
    # We then pipe this to 'source 3<&0’.
    #
    # You might expect that the dup2(3, stdin) should overwrite stdin,
    # and therefore prevent 'source' from reading the piped-in block. This doesn't happen
    # because when you pipe to a builtin, we don't overwrite stdin with the read end
    # of the block; instead we set a separate fd in a variable 'builtin_stdin', which is
    # what it reads from. So builtins are magic in that, in pipes, their stdin
    # is not fd 0.
    #
    # ‘source’ does not apply the redirections to itself. Instead it saves them and passes
    # them as block-level redirections to parser.eval(). Ultimately the eval’d code sees
    # the following redirections (in the following order):
    #    dup2 0 -> 3
    #    dup2 pipe -> 0
    #    dup2 3 -> 0
    # where the pipe is the pipe we get from piping ‘echo’ to ‘source’. Thus the redirection
    # effectively makes stdin fd0, instead of the thing that was piped to ‘source’
    echo "begin; $argv "\n" ;end <&3 3<&-" | source 3<&0
    set -l res $status

    status --job-control $mode
    return $res
end
```

**Comment by faho** [comment]
Okay. You have an old `eval` function left over from an upgrade.

Remove /usr/local/share/fish/functions/eval.fish to use the builtin.

**Comment by wjljack** [comment]
and use bash or zsh with `fzf-file-widget` is ok

**Comment by wjljack** [comment]
> 

I will try

**Comment by wjljack** [comment]
after delete `/usr/local/share/fish/functions/eval.fish` , everything is ok , awesome !!!
I like fish very mush,with many features than zsh(oh-my-zsh is slow and large)

**Comment by faho** [comment]
Okay, problem was an incomplete upgrade, closing!

**Comment by faho** [comment]
Okay, I checked and it turns out `eval` can no longer be safely shadowed by a function. Since we made it a builtin, it no longer starts a block, so `eval set -l` works. As soon as you replace it with a function that's broken.

So I turned the name into a reserved keyword, which means future fish versions will simply ignore an "eval.fish" file, and if you explicitly told them to `source` it would print an error. So this specific failure case no longer exists.
